# load data (1)

In [19]:
import pandas as pd
import numpy as np

In [27]:
df = pd.read_csv('https://raw.githubusercontent.com/msfrv/interviews/main/prices.csv')
df.head()

stock        date      price
0     A  2018-01-01  42.910205
1     A  2018-01-02  46.468834
2     A  2018-01-03  71.158140
3     A  2018-01-04  43.746151
4     A  2018-01-05  43.036749

## pivot table
Convert the dataframe such that the rows are date, columns are stocks.
The first few rows should be like:

```
                   A          B          C          D          E          F   ...
2018-01-01  42.910205  29.145580  66.738553  55.920965  90.381230  50.161381  ... 
2018-01-02  46.468834  29.513768  62.218892  56.656698  88.731589  47.227976  ...
2018-01-03  71.158140  29.877393  66.628797  55.212968  90.360738  48.242614  ... 
2018-01-04  43.746151  29.774971  62.862005  58.668346  93.931583  49.503737  ... 
2018-01-05  43.036749  29.691612  65.681785  55.069295  90.879364  48.997726  ...
```



In [28]:
# code here


## find outliers
find datapoints (stock+date) which are >2 sigmas (standard deviation) away from their means.

Show their z-scores and the prices.

The output should be like:

```
 stock        date   z-score      price
     A  2018-01-03  11.047592  71.158140
     B  2018-07-27   7.976222  32.058381
     E  2018-02-15  -4.397496  82.856551
......
```
note: z-score is $(x-\mu)/\sigma$


In [26]:
# code here


## remove outliers
for each of the outliers found above, replace their prices by interpolating between the values before and after.

For example, for the first outlier (price of A at 2018-01-03):

```
stock  date           prices
A      2018-01-02     46.468834
       2018-01-03     71.158140
       2018-01-04     43.746151
```
We will replace the price of A at 2018-01-03 by: $(46.47+43.75)/2=45.11$

Output the list of replacements used:
```
stock  date      
A      2018-01-03    45.107493
B      2018-07-27    29.274156
......
```


In [29]:
# code here


The prices should now be like this :
```
                    A          B          C          D          E                                                                       
2018-01-01  42.910205  29.145580  66.738553  55.920965  90.381230 ...  
2018-01-02  46.468834  29.513768  62.218892  56.656698  88.731589 ...  
2018-01-03  45.107493  29.877393  66.628797  55.212968  90.360738 ...  
2018-01-04  43.746151  29.774971  62.862005  58.668346  93.931583 ... 
...
```  

# load weights
This file contains the weights of those 10 stocks in 20 ETFs.
Each row is guaranteed to sum to 1.0

In [32]:
w = pd.read_csv('https://raw.githubusercontent.com/msfrv/interviews/main/weights.csv',index_col=0)
w.head()

A         B         C         D         E         F         G  \
ETF0  0.032161  0.160087  0.155371  0.017495  0.158185  0.053476  0.221998   
ETF1  0.020030  0.078445  0.149019  0.091011  0.006436  0.190682  0.148907   
ETF2  0.196113  0.181189  0.095938  0.086834  0.006523  0.083930  0.099427   
ETF3  0.048181  0.163308  0.053828  0.092525  0.107521  0.157659  0.005138   
ETF4  0.120828  0.127946  0.200062  0.122529  0.108332  0.153985  0.066616   

             H         I         J  
ETF0  0.087146  0.113267  0.000814  
ETF1  0.161568  0.087372  0.066530  
ETF2  0.115733  0.031817  0.102496  
ETF3  0.162741  0.101255  0.107844  
ETF4  0.050126  0.025478  0.024097

Please compute the *recomposed returns* of the ETFs by:

$ r_{i,d} = \sum_j w_{i,j} (p_{j,d}/p_{j,d-1}-1)$

where 
* $r_{i,d}$ is the recomposed return of ETF$i$ on date $d$
* $w_{i,j}$ is the weight of the $j^{th}$ stock in ETF$i$
* $p_{j,d}$ is the price of the $j^{th}$ stock on date $d$

The result should be a 260x20 dataframe

In [35]:
# code here